internal/terraform/eval_variable.go

```diff
@@ -90,8 +90,11 @@ func prepareFinalInputVariableValue(addr addrs.AbsInputVariableInstance, raw *In
 		given = defaultVal // must be set, because we checked above that the variable isn't required
 	}
 
-	// Apply defaults from the variable's type constraint to the given value
-	if cfg.TypeDefaults != nil {
+	// Apply defaults from the variable's type constraint to the given value,
+	// unless the given value is null. We do not apply defaults to top-level
+	// null values, as doing so could prevent assigning null to a nullable
+	// variable.
+	if cfg.TypeDefaults != nil && !given.IsNull() {
```

Comment:
The direct `null` assignment worked because it was a null value of dynamic type. This caused [the defaults implementation to shortcut return](https://github.com/hashicorp/hcl/blob/85e45c55d72c28821cad8abc2f016c53b833b87c/ext/typeexpr/defaults.go#L72-L79), which was accidentally desirable.

In contrast, we type-convert default values to the variable's type constraint on config load. So when we get here with a null value of an object type, we'd then apply the defaults, creating a non-null object.